### PR TITLE
Add find tag command

### LINKED ISSUE #255: As a user, I can search for tags
so that I can find certain types of tags by a partial or full match.

### LINKED ISSUE #256: Add find tag command
allows checking by partial match of the tag's name, as well as filtering by whether they are super tags, regular tags, or both.

### PR DESCRIPTION
**Pull request by fyshhh**
Closes #255 and #256

### CODE REVIEW

File: src/main/java/seedu/address/logic/parser/tags/FindTagCommandParser.java [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+     * keyword, type of tag (super or regular), or both, depending on user input.
+     * @throws ParseException if the user input does not conform to the expected format
+     */
+    public FindTagCommand parse(String args) throws ParseException {
+        assert args != null;
+
+        ArgumentMultimap argumentMultimap =
+                ArgumentTokenizer.tokenize(args, PREFIX_TAG, PREFIX_SUPERTAG_ONLY);
+
+        Optional<NameContainsKeywordsPredicate> findPredicate = Optional.empty();
```
Comment (bangyiwu) [inline src/main/java/seedu/address/logic/parser/tags/FindTagCommandParser.java:37]: Is there an exception being thrown for st/ values of more than 1 or less than 0?

Comment (fyshhh) [reply src/main/java/seedu/address/logic/parser/tags/FindTagCommandParser.java:37]: yep, this is handled in the `BooleanInput` class.

Screenshot 2020-11-07 at 3 16 46 PM

File: src/main/java/seedu/address/logic/commands/tags/FindTagCommand.java [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+
+    public static final String MESSAGE_NO_TAGS_FOUND = "No tag assignments were found!";
+
+    public static final String MESSAGE_TAGS_FOUND = "Listed all tags and contacts directly under these tags:";
+
+    public static final String MESSAGE_INVALID_SEARCH_FIELD = "Tag queries should be alphanumeric with no spaces.";
+
+    public static final String INDICATOR_SUPERTAG = " (supertag)";
+
+    public static final String INDICATOR_NO_CONTACTS_TAGGED = "no contacts tagged";
```
Comment (chan-j-d) [inline src/main/java/seedu/address/logic/commands/tags/FindTagCommand.java]: Besides the `MESSAGE_INVALID_SEARCH_FIELD`, the rest seem to overlap with the `ViewTagCommand`. can consider abstracting these out to a general `TagUtil` class.

File: src/main/java/seedu/address/logic/commands/tags/FindTagCommand.java [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+     * Returns a string that combines the {@code Person} set by commas, then adds curly braces.
+     * If the set is empty, returns {@code messageIfNoneFound}.
+     */
+    protected static String parsePersonSetIntoString(Set<Person> set, String messageIfNoneFound) {
+        return set.stream()
+                .map(t -> t.getName().toString())
+                .reduce((s1, s2) -> s1 + ", " + s2)
+                .map(string -> "{ " + string + " }")
+                .orElse(messageIfNoneFound);
+    }
```
Comment (chan-j-d) [inline src/main/java/seedu/address/logic/commands/tags/FindTagCommand.java]: I believe this method is also shared in `ViewTagCommand`. Can consider abstracting it out as well

File: src/test/data/JsonSerializableTagTreeTest/sampleValidTagTree.json [hunk 4]
```diff
@@ -1,5 +1,6 @@
 {
   "tagSubTagMap" : {
+    "architecture" : [ "architecturemod" ],
```
Comment (chan-j-d) [inline src/test/data/JsonSerializableTagTreeTest/sampleValidTagTree.json:3]: Pretty daring to change the sample test tree. But I guess maybe the original one might not have been comprehensive enough.

**Review by chan-j-d — CHANGES_REQUESTED** [review]
Looks really good! Just some minor overlaps in code with `ViewTagCommand` as I thought the methods in there would be standalone. Could try to abstract them out.

File: src/main/java/seedu/address/logic/commands/tags/FindTagCommand.java (hunk 2 above)
Comment (fyshhh) [reply src/main/java/seedu/address/logic/commands/tags/FindTagCommand.java]: good catch! i will do so.

File: src/test/data/JsonSerializableTagTreeTest/sampleValidTagTree.json (hunk 4 above)
Comment (fyshhh) [reply src/test/data/JsonSerializableTagTreeTest/sampleValidTagTree.json:3]: haha, i ran the tests again to ensure nothing else was affected.

File: src/main/java/seedu/address/commons/core/booleaninput/BooleanInput.java [hunk 5]
```diff
@@ -11,15 +11,15 @@
...
 
     private static final BooleanInput BOOLEAN_TRUE = new BooleanInput(true);
     private static final BooleanInput BOOLEAN_FALSE = new BooleanInput(false);
 
-    public static final String MESSAGE_CONSTRAINTS = "A boolean indicator must be one of the following: "
-            + String.join(",", SET_TRUE_STRINGS) + " (true) "
-            + String.join(",", SET_FALSE_STRINGS) + " (false)";
+    public static final String MESSAGE_CONSTRAINTS = "A boolean indicator must be one of the following:\n"
+            + "[true]: " + String.join(", ", SET_TRUE_STRINGS) + "\n"
+            + "[false]: " + String.join(", ", SET_FALSE_STRINGS);
```
Comment (chan-j-d) [inline src/main/java/seedu/address/commons/core/booleaninput/BooleanInput.java:22]: great catch!

**Review by chan-j-d — APPROVED** [review]
LGTM!

**Pull request merged**